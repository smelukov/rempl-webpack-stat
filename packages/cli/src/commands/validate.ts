import path from 'path';
import { Argv } from 'yargs';
import Validator, { ValidationResult } from '@statoscope/stats-validator';
import ConsoleReporter from '@statoscope/stats-validator-reporter-console';
import legacyWebpackStatsValidator from './legacyWebpackValidator';

export default function (yargs: Argv): Argv {
  return yargs.command(
    'validate [validator] [config] [input]',
    `[BETA] Validate one or more JSON stats
Examples:
Single stats: validate path/to/validator.js path/to/stats.json
Multiple stats: generate path/to/validator.js --input path/to/stats-1.json path/to/stats-2.json`,
    (yargs) => {
      return yargs
        .positional('validator', {
          describe: 'path to validator script',
          alias: 'v',
          type: 'string',
        })
        .positional('config', {
          describe: 'path to statoscope config',
          alias: 'c',
          type: 'string',
        })
        .positional('input', {
          describe: 'path to a stats.json',
          alias: 'i',
          type: 'string',
        })
        .option('warn-as-error', {
          type: 'boolean',
          describe: 'Treat warnings as errors',
          alias: 'w',
        })
        .array('input')
        .demandOption(['input']);
    },
    async (argv) => {
      let result: ValidationResult | undefined;

      // todo statoscope 6: remove this or use as prepare + validator
      if (argv.validator) {
        result = await legacyWebpackStatsValidator(
          path.relative(process.cwd(), argv.validator),
          argv.input,
          {
            warnAsError: argv['warn-as-error'],
          }
        );
      } else {
        // todo resole nearest config
        const configPath = path.resolve(argv.config);
        const validator = new Validator(configPath);

        result = await validator.validate(argv.input.map((file) => path.resolve(file)));
      }

      const reporter = new ConsoleReporter();
      await reporter.run(result, { warnAsError: argv['warn-as-error'] });
    }
  );
}
