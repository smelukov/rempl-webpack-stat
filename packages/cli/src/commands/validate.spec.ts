import path from 'path';
import yargs from 'yargs';
import validate from './validate';

const validatorFixturesJoraQuery = [
  '../../../../test/fixtures/cli/validate/ok.jora',
  '../../../../test/fixtures/cli/validate/fail.jora',
  '../../../../test/fixtures/cli/validate/rules.jora',
].map((filename) => ({
  name: path.basename(filename),
  path: path.resolve(__dirname, filename),
}));

const validatorFixturesJSJoraQuery = [
  '../../../../test/fixtures/cli/validate/jora-ok.js',
  '../../../../test/fixtures/cli/validate/jora-fail.js',
].map((filename) => ({
  name: path.basename(filename),
  path: path.resolve(__dirname, filename),
}));

const inputFixturePath = [
  '../../../../test/bundles/simple/stats-dev.json',
  '../../../../test/bundles/v4/simple/stats-dev.json',
].map((filepath) => path.resolve(__dirname, filepath));

// const consoleLog = console.log.bind(console);
const output: string[][] = [];
let consoleLogSpy: jest.SpyInstance<void, string[]> | null = null;

beforeEach(() => {
  consoleLogSpy = jest
    .spyOn(global.console, 'log')
    .mockImplementation((...args: string[]) => {
      output.push(args.map((o) => String(o).replace(process.cwd(), '<pwd>')));
    });
});

afterEach(() => {
  consoleLogSpy?.mockRestore();
  output.length = 0;
});

function getOutput(): string[][] {
  return output;
}

describe('validator types', () => {
  test.each(validatorFixturesJoraQuery)('raw jora-query $name', async (item) => {
    let y = yargs(['validate', '--validator', item.path, '--input', ...inputFixturePath]);

    y = validate(y);
    y.fail((_, error) => {
      console.error(error);
    });

    await y.argv;

    expect(getOutput()).toMatchSnapshot();
  });

  test.each(validatorFixturesJSJoraQuery)('jora-query from js $name', async (item) => {
    let y = yargs(['validate', '--validator', item.path, '--input', ...inputFixturePath]);

    y = validate(y);
    y.fail((_, error) => {
      console.error(error);
    });

    await y.argv;

    expect(getOutput()).toMatchSnapshot();
  });
});
