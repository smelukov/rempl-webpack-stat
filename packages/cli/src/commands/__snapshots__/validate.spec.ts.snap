// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validator types jora-query from js jora-fail.js 1`] = `
Array [
  Array [
    "[4munknown file[24m",
  ],
  Array [
    "[36munknown compilation[39m",
  ],
  Array [
    "[31merror[39m  Fail  [36m<pwd>/test/fixtures/cli/validate/jora-fail.js[39m",
  ],
  Array [
    "",
  ],
  Array [
    "[33m1 problems (1 errors, 0 warnings)[39m",
  ],
]
`;

exports[`validator types jora-query from js jora-ok.js 1`] = `Array []`;

exports[`validator types raw jora-query fail.jora 1`] = `
Array [
  Array [
    "[4munknown file[24m",
  ],
  Array [
    "[36munknown compilation[39m",
  ],
  Array [
    "[31merror[39m  Fail  [36m<pwd>/test/fixtures/cli/validate/fail.jora[39m",
  ],
  Array [
    "",
  ],
  Array [
    "[33m1 problems (1 errors, 0 warnings)[39m",
  ],
]
`;

exports[`validator types raw jora-query ok.jora 1`] = `Array []`;

exports[`validator types raw jora-query rules.jora 1`] = `
Array [
  Array [
    "[4munknown file[24m",
  ],
  Array [
    "[36munknown compilation[39m",
  ],
  Array [
    "[0minfo[0m  Auto info message  [36m<pwd>/test/fixtures/cli/validate/rules.jora[39m",
  ],
  Array [
    "[0minfo[0m  Info message without assert  [36m<pwd>/test/fixtures/cli/validate/rules.jora[39m",
  ],
  Array [
    "[0minfo[0m  Info with assert  [36m<pwd>/test/fixtures/cli/validate/rules.jora[39m",
  ],
  Array [
    "[33mwarn[39m  Warn without assert  [36m<pwd>/test/fixtures/cli/validate/rules.jora[39m",
  ],
  Array [
    "[33mwarn[39m  Warn with assert  [36m<pwd>/test/fixtures/cli/validate/rules.jora[39m",
  ],
  Array [
    "",
  ],
  Array [
    "[4m<pwd>/test/bundles/simple/stats-dev.json[24m",
  ],
  Array [
    "[36munknown compilation[39m",
  ],
  Array [
    "[0minfo[0m  Auto info for file  [36m<pwd>/test/fixtures/cli/validate/rules.jora[39m",
  ],
  Array [
    "[0minfo[0m  Info for file  [36m<pwd>/test/fixtures/cli/validate/rules.jora[39m",
  ],
  Array [
    "[33mwarn[39m  Warn for file  [36m<pwd>/test/fixtures/cli/validate/rules.jora[39m",
  ],
  Array [
    "[31merror[39m  Error for file  [36m<pwd>/test/fixtures/cli/validate/rules.jora[39m",
  ],
  Array [
    "[36med4bbcc136a84609ae9f[39m",
  ],
  Array [
    "[31merror[39m  Something wrong  [36m<pwd>/test/fixtures/cli/validate/rules.jora[39m",
  ],
  Array [
    "",
  ],
  Array [
    "[33m5 problems (2 errors, 3 warnings)[39m",
  ],
]
`;
